File: Folicon_Native/Modules/DArt.vb
Imports System.Configuration
Imports Newtonsoft.Json

Namespace Modules
    Public Module DArt
        Private Async Function GenerateNewAccessToken() As Task(of string)
            Dim clientAccessToken = ""
            Dim url = "https://www.deviantart.com/oauth2/token?client_id=" & ClientIdDArt &
                      "&client_secret=" & ClientSecretDArt &
                      "&grant_type=client_credentials"
            Using response = Await HttpC.GetAsync(New Uri(url))
                Dim jsonData = Await response.Content.ReadAsStringAsync()
                dim tokenResponse = JsonConvert.DeserializeObject (Of DArtTokenResponse)(jsonData)
                dim config As Configuration =
                        ConfigurationManager.OpenExeConfiguration(ConfigurationUserLevel.None)
                clientAccessToken = tokenResponse.AccessToken
                config.AppSettings.Settings("Token").Value = tokenResponse.AccessToken
                config.Save(ConfigurationSaveMode.Modified)
                ConfigurationManager.RefreshSection("appSettings")
            End Using
            Return clientAccessToken
        End Function

        public Async function GetClientAccessTokenAsync() As Task(of string)
            Dim clientAccessToken As String = ConfigurationManager.AppSettings.Get("Token")
            If Not string.IsNullOrEmpty(clientAccessToken)
                If not Await IsTokenValidAsync(clientAccessToken)
                    clientAccessToken = Await GenerateNewAccessToken()
                End If
            Else
                clientAccessToken = Await GenerateNewAccessToken()
            End If
            Return clientAccessToken
        end Function

        public Async function IsTokenValidAsync(clientAccessToken As string) As Task(of Boolean)
            Dim url = "https://www.deviantart.com/api/v1/oauth2/placebo?access_token=" & clientAccessToken
            Dim tokenResponse As DArtTokenResponse
            using response = Await httpC.GetAsync(New Uri(url))
                Dim jsonData = Await response.Content.ReadAsStringAsync()
                tokenResponse = JsonConvert.DeserializeObject (Of DArtTokenResponse)(jsonData)
            End Using
            If tokenResponse.Status = "success"
                Return true
            Else
                Return false
            End If
        End function

        public Async Function Browse(accessToken As string, query As string,Optional offset As Integer=0) As Task(Of DArtBrowseResult)
            Dim url ="https://www.deviantart.com/api/v1/oauth2/browse/popular?timerange=alltime&offset=" & offset & "&category_path=customization%2ficons%2fos%2fwin&q=" &
                    query & " folder icon" & "&limit=20&access_token=" & accessToken
            Dim result As DArtBrowseResult
            Using response = Await HttpC.GetAsync(New Uri(url))
                Dim jsonData = Await response.Content.ReadAsStringAsync()
                dim serializerSettings = new JsonSerializerSettings _
                        With{ .NullValueHandling = NullValueHandling.Ignore }
                result = JsonConvert.DeserializeObject (Of DArtBrowseResult)(jsonData, serializerSettings)
            End Using
            Return result
        End Function
    End Module
End Namespace

File: FoliCon/Modules/DArt.cs
namespace FoliCon.Modules
{
    public class DArt : BindableBase
    {
        private string _clientAccessToken;
        private string _clientSecret;
        private string _clientId;

        public string ClientId
        {
            get => _clientId;
            set => SetProperty(ref _clientId, value);
        }

        public string ClientSecret
        {
            get => _clientSecret;
            set => SetProperty(ref _clientSecret, value);
        }

        public string ClientAccessToken
        {
            get => _clientAccessToken;
            set => SetProperty(ref _clientAccessToken, value);
        }

        public DArt(string clientSecret, string clientId)
        {
            Services.Tracker.Configure<DArt>()
                .Property(p => p.ClientAccessToken)
                .PersistOn(nameof(PropertyChanged));
            ClientSecret = clientSecret;
            ClientId = clientId;
            Services.Tracker.Track(this);
            GetClientAccessTokenAsync();
        }

        public async void GetClientAccessTokenAsync()
        {
            if (!string.IsNullOrEmpty(ClientAccessToken))
            {
                if (!await IsTokenValidAsync(ClientAccessToken))
                {
                    ClientAccessToken = await GenerateNewAccessToken();
                }
            }
            else
            {
                ClientAccessToken = await GenerateNewAccessToken();
            }
        }

        public static async Task<bool> IsTokenValidAsync(string clientAccessToken)
        {
            var url = "https://www.deviantart.com/api/v1/oauth2/placebo?access_token=" + clientAccessToken;
            using var response = await Services.HttpC.GetAsync(new Uri(url));
            var jsonData = await response.Content.ReadAsStringAsync();
            var tokenResponse = JsonConvert.DeserializeObject<DArtTokenResponse>(jsonData);

            return tokenResponse.Status == "success";
        }

        private async Task<string> GenerateNewAccessToken()
        {
            var url = "https://www.deviantart.com/oauth2/token?client_id=" + ClientId +
                      "&client_secret=" + ClientSecret +
                      "&grant_type=client_credentials";
            using var response = await Services.HttpC.GetAsync(new Uri(url));
            var jsonData = await response.Content.ReadAsStringAsync();
            var tokenResponse = JsonConvert.DeserializeObject<DArtTokenResponse>(jsonData);
            var clientAccessToken = tokenResponse.AccessToken;
            return clientAccessToken;
        }

        public async Task<DArtBrowseResult> Browse(string query, int offset = 0)
        {
            GetClientAccessTokenAsync();
            var url = "https://www.deviantart.com/api/v1/oauth2/browse/popular?timerange=alltime&offset=" + offset +
                      "&q=" + query + " folder icon" +
                      "&limit=20&access_token=" + ClientAccessToken;
            using var response = await Services.HttpC.GetAsync(new Uri(url));
            var jsonData = await response.Content.ReadAsStringAsync();
            var serializerSettings = new JsonSerializerSettings { NullValueHandling = NullValueHandling.Ignore };
            var result = JsonConvert.DeserializeObject<DArtBrowseResult>(jsonData, serializerSettings);
            return result;
        }
    }
}